beacon-chain/blockchain/core.go

```diff
@@ -319,59 +303,105 @@ func (b *BeaconChain) saveBlock(block *types.Block) error {
 
-		// Verify attester bitfields matches crystallized state's prev computed bitfield.
-		if err := b.validateAttesterBitfields(attestation, attesterIndices); err != nil {
-			return err
-		}
+	// Verify attester bitfields matches crystallized state's prev computed bitfield.
+	if err := b.validateAttesterBitfields(attestation, attesterIndices); err != nil {
+		return err
+	}
 
-		// TODO: Generate validators aggregated pub key.
+	// TODO: Generate validators aggregated pub key.
 
-		// Hash parentHashes + shardID + slotNumber + shardBlockHash into a message to use to
-		// to verify with aggregated public key and aggregated attestation signature.
-		msg := make([]byte, binary.MaxVarintLen64)
-		var signedHashesStr []byte
-		for _, parentHash := range parentHashes {
-			signedHashesStr = append(signedHashesStr, parentHash.Bytes()...)
-			signedHashesStr = append(signedHashesStr, byte(' '))
-		}
-		binary.PutUvarint(msg, attestation.Slot%params.CycleLength)
-		msg = append(msg, signedHashesStr...)
-		binary.PutUvarint(msg, attestation.ShardId)
-		msg = append(msg, attestation.ShardBlockHash...)
+	// Hash parentHashes + shardID + slotNumber + shardBlockHash into a message to use to
+	// to verify with aggregated public key and aggregated attestation signature.
+	msg := make([]byte, binary.MaxVarintLen64)
+	var signedHashesStr []byte
+	for _, parentHash := range parentHashes {
+		signedHashesStr = append(signedHashesStr, parentHash.Bytes()...)
+		signedHashesStr = append(signedHashesStr, byte(' '))
+	}
+	binary.PutUvarint(msg, attestation.Slot%params.CycleLength)
+	msg = append(msg, signedHashesStr...)
+	binary.PutUvarint(msg, attestation.ShardId)
+	msg = append(msg, attestation.ShardBlockHash...)
 
-		msgHash := blake2b.Sum512(msg)
+	msgHash := blake2b.Sum512(msg)
 
-		log.Debugf("Attestation message for shard: %v, slot %v, block hash %v is: %v",
-			attestation.ShardId, attestation.Slot, attestation.ShardBlockHash, msgHash)
+	log.Debugf("Attestation message for shard: %v, slot %v, block hash %v is: %v",
+		attestation.ShardId, attestation.Slot, attestation.ShardBlockHash, msgHash)
 
-		// TODO: Verify msgHash against aggregated pub key and aggregated signature.
-	}
+	// TODO: Verify msgHash against aggregated pub key and aggregated signature.
 	return nil
 }
 
+// calculateBlockVoteCache calculates and updates active state's block vote cache.
+func (b *BeaconChain) calculateBlockVoteCache(attestationIndex int, block *types.Block) (map[*common.Hash]*types.VoteCache, error) {
+	attestation := block.Attestations()[attestationIndex]
+	newVoteCache := b.ActiveState().GetBlockVoteCache()
+	parentHashes := b.getSignedParentHashes(block, attestation)
+	attesterIndices, err := b.getAttesterIndices(attestation)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, h := range parentHashes {
+		// Skip calculating for this hash if the hash is part of oblique parent hashes.
+		var skip bool
+		for _, obliqueParentHash := range attestation.ObliqueParentHashes {
```

Comment:
Wouldn't this be more concise if `getSignedParentHashes` didn't return any of the oblique parent hashes? Note that `getSignedParentHashes` is appending the oblique parent hashes to the end of itso wn list. This is adding unnecessary complexity.

Reply:
I think in the spec, `getSignedParentHashes` means all the parent hashes (including now oblique) that you have signed which is still important to know. For example you signed A->B->C->D->E and there is A->B->C`->D`->E`, but I still want to get A->B->C->D. Does that make sense?

Reply:
I need to sit on this a little more, but that shouldn't block this PR, since it's not blocking anything.